### PR TITLE
Refactor RuleEngine states

### PR DESCRIPTION
**Pull request by zigen**
prev: #331 
* applied_rules : RuleEngine -> QubitRecord

---
This change is [Reviewable](https://reviewable.io/reviews/sfc-aqua/quisp/332)

### CODE REVIEW

**Review by Naphann — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/sfc-aqua/quisp/332)* status: 0 of 3 files reviewed, 1 unresolved discussion (waiting on @Naphann and @zigen)

___
*[quisp/modules/QRSA/RuleEngine/RuleEngine.cc, line 1205 at r1](https://reviewable.io/reviews/sfc-aqua/quisp/332#-MuoEGnz4SLssEw4-Fk-:-MuoEGo-DYh5e2f6Cf6O:bxqnjrr) ([raw file](https://github.com/sfc-aqua/quisp/blob/fb5cb0fca19b7f91a3ef2fc92b6b176d229216a9/quisp/modules/QRSA/RuleEngine/RuleEngine.cc#L1205)):*
> ```cpp
>   realtime_controller->ReInitialize_StationaryQubit(qubit_record, false);
>   qubit_record->setBusy(false);
>   if (qubit_record->isAllocated()) {
> ```

do we need this check? If in the end we want it to always be `false`?

**Review by Naphann — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/sfc-aqua/quisp/332)* status: 0 of 3 files reviewed, 1 unresolved discussion (waiting on @Naphann and @zigen)

___
*[quisp/modules/QRSA/RuleEngine/RuleEngine.cc, line 1205 at r1](https://reviewable.io/reviews/sfc-aqua/quisp/332#-MuoEGnz4SLssEw4-Fk-:-MuoFsilCK1gX_cztQzi:bjr08ho) ([raw file](https://github.com/sfc-aqua/quisp/blob/fb5cb0fca19b7f91a3ef2fc92b6b176d229216a9/quisp/modules/QRSA/RuleEngine/RuleEngine.cc#L1205)):*
<details><summary><i>Previously, Naphann (Naphan Benchasattabuse) wrote…</i></summary><blockquote>

do we need this check? If in the end we want it to always be `false`?
</blockquote></details>

I mean, should setBusy care about the current state of the qubit?

**Review by zigen — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/sfc-aqua/quisp/332)* status: 0 of 3 files reviewed, 1 unresolved discussion (waiting on @Naphann)

___
*[quisp/modules/QRSA/RuleEngine/RuleEngine.cc, line 1205 at r1](https://reviewable.io/reviews/sfc-aqua/quisp/332#-MuoEGnz4SLssEw4-Fk-:-MusjVC63e-CEgGzBEwc:bfsmx3x) ([raw file](https://github.com/sfc-aqua/quisp/blob/fb5cb0fca19b7f91a3ef2fc92b6b176d229216a9/quisp/modules/QRSA/RuleEngine/RuleEngine.cc#L1205)):*
<details><summary><i>Previously, Naphann (Naphan Benchasattabuse) wrote…</i></summary><blockquote>

I mean, should setBusy care about the current state of the qubit? 
</blockquote></details>

for now, `setBusy` cares about the current state of the qubit. 
basically, I copied that from the original implementation.
https://github.com/sfc-aqua/quisp/blob/ee8873aa074b13687bcfaa703462d67811ab587b/quisp/modules/RuleEngine.cc#L873
but it might be useless now. let's remove it later.

**Review by Naphann — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/sfc-aqua/quisp/332)* status: 0 of 3 files reviewed, 1 unresolved discussion (waiting on @Naphann and @zigen)

___
*[quisp/modules/QRSA/RuleEngine/RuleEngine.cc, line 1205 at r1](https://reviewable.io/reviews/sfc-aqua/quisp/332#-MuoEGnz4SLssEw4-Fk-:-Musm9WD7AoKlacvQjLz:b-l8u4ca) ([raw file](https://github.com/sfc-aqua/quisp/blob/fb5cb0fca19b7f91a3ef2fc92b6b176d229216a9/quisp/modules/QRSA/RuleEngine/RuleEngine.cc#L1205)):*
<details><summary><i>Previously, zigen (Kentaro "zigen" Teramoto) wrote…</i></summary><blockquote>

for now, `setBusy` cares about the current state of the qubit. 
basically, I copied that from the original implementation.
https://github.com/sfc-aqua/quisp/blob/ee8873aa074b13687bcfaa703462d67811ab587b/quisp/modules/RuleEngine.cc#L873
but it might be useless now. let's remove it later.
</blockquote></details>

Normally when a method can throw due to something like this I'd name it try in front. In this case it would be`trySetBusy`

**Review by zigen — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/sfc-aqua/quisp/332)* status: 0 of 3 files reviewed, 1 unresolved discussion (waiting on @Naphann)

___
*[quisp/modules/QRSA/RuleEngine/RuleEngine.cc, line 1205 at r1](https://reviewable.io/reviews/sfc-aqua/quisp/332#-MuoEGnz4SLssEw4-Fk-:-MusyLfk29EIITy3EZx2:b-pz9cmq) ([raw file](https://github.com/sfc-aqua/quisp/blob/fb5cb0fca19b7f91a3ef2fc92b6b176d229216a9/quisp/modules/QRSA/RuleEngine/RuleEngine.cc#L1205)):*
<details><summary><i>Previously, Naphann (Naphan Benchasattabuse) wrote…</i></summary><blockquote>

Normally when a method can throw due to something like this I'd name it try in front. In this case it would be`trySetBusy`
</blockquote></details>

that's a good idea!

**Review by Naphann — APPROVED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sfc-aqua/quisp/332)* status: 0 of 3 files reviewed, 1 unresolved discussion (waiting on @Naphann and @zigen)

___
*[quisp/modules/QRSA/RuleEngine/RuleEngine.cc, line 1205 at r1](https://reviewable.io/reviews/sfc-aqua/quisp/332#-MuoEGnz4SLssEw4-Fk-:-Mut_7s8Cx7gUM9DxBmY:b-fn2d1c) ([raw file](https://github.com/sfc-aqua/quisp/blob/fb5cb0fca19b7f91a3ef2fc92b6b176d229216a9/quisp/modules/QRSA/RuleEngine/RuleEngine.cc#L1205)):*
<details><summary><i>Previously, zigen (Kentaro "zigen" Teramoto) wrote…</i></summary><blockquote>

that's a good idea!
</blockquote></details>

I think I'll approve it for and let's change that later

**Review by zigen — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/sfc-aqua/quisp/332)* status: 0 of 3 files reviewed, all discussions resolved (waiting on @Naphann)

___
*[quisp/modules/QRSA/RuleEngine/RuleEngine.cc, line 1205 at r1](https://reviewable.io/reviews/sfc-aqua/quisp/332#-MuoEGnz4SLssEw4-Fk-:-MuxWf-29tFSPL1-9AEc:bks6w1j) ([raw file](https://github.com/sfc-aqua/quisp/blob/fb5cb0fca19b7f91a3ef2fc92b6b176d229216a9/quisp/modules/QRSA/RuleEngine/RuleEngine.cc#L1205)):*
<details><summary><i>Previously, Naphann (Naphan Benchasattabuse) wrote…</i></summary><blockquote>

I think I'll approve it for and let's change that later
</blockquote></details>

okay, let's refactor it later. thanks!

**Comment by zigen** [comment]
@Naphann thanks for reviewing, if it's okay, just LGTM it plz.

**Review by Naphann — COMMENTED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/sfc-aqua/quisp/332#-:-MuxXPmmEmu3oI4cYEyW:bnfp4nl)* status: 0 of 3 files reviewed, all discussions resolved (waiting on @Naphann)

**Review by Naphann — APPROVED** [review]
> Reviewed 3 of 3 files at r1.
> *[Reviewable](https://reviewable.io/reviews/sfc-aqua/quisp/332)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @zigen)
